File: app/src/main/java/com/example/yelpbusinessexample/data/api/NetworkService.java
package com.example.yelpbusinessexample.data.api;

import okhttp3.OkHttpClient;
import okhttp3.logging.HttpLoggingInterceptor;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

public class NetworkService {

   private static final String BASE_URL = " https://api.yelp.com/v3/";

   private final YelpServiceApi yelpServiceApi;
   private static NetworkService service;

   public NetworkService() {
      HttpLoggingInterceptor logging = new HttpLoggingInterceptor();
      logging.setLevel(HttpLoggingInterceptor.Level.BASIC);
      OkHttpClient client = new OkHttpClient.Builder()
              .addInterceptor(logging)
              .build();

      Retrofit retrofit = new Retrofit.Builder().baseUrl(BASE_URL).client(client).addConverterFactory(GsonConverterFactory.create()).build();

      yelpServiceApi = retrofit.create(YelpServiceApi.class);
   }

   public static NetworkService getInstance() {
      if (service == null) {
         synchronized (NetworkService.class) {
            if (service == null) {
               service = new NetworkService();
            }
         }
      }
      return service;
   }

   public YelpServiceApi getYelpServiceApi() {
      return yelpServiceApi;
   }

}

File: app/src/main/java/com/example/yelpbusinessexample/data/api/NetworkService.kt
package com.example.yelpbusinessexample.data.api

import okhttp3.logging.HttpLoggingInterceptor
import okhttp3.OkHttpClient
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory

class NetworkService {
    val yelpServiceApi: YelpServiceApi

    companion object {
        private const val BASE_URL = " https://api.yelp.com/v3/"
        private var service: NetworkService? = null

        @JvmStatic
        val instance: NetworkService?
            get() {
                if (service == null) {
                    synchronized(NetworkService::class.java) {
                        if (service == null) {
                            service = NetworkService()
                        }
                    }
                }
                return service
            }
    }

    init {
        val logging = HttpLoggingInterceptor()
        logging.setLevel(HttpLoggingInterceptor.Level.BASIC)
        val client: OkHttpClient = OkHttpClient.Builder().apply {
            addInterceptor(logging)
        }.build()

        val retrofit = Retrofit.Builder().apply {
            baseUrl(BASE_URL)
            client(client)
            addConverterFactory(GsonConverterFactory.create())
        }.build()

        yelpServiceApi = retrofit.create(YelpServiceApi::class.java)
    }
}